blocks/certifier.go

```diff
@@ -203,27 +214,50 @@ func (c *Certifier) RegisterForCert(ctx context.Context, lid types.LayerID, bid
 
 	c.mu.Lock()
 	defer c.mu.Unlock()
-	c.createIfNeeded(lid, bid)
-	c.certifyMsgs[lid][bid].registered = true
-	return c.tryGenCert(ctx, logger, lid, bid)
+	info := c.createIfNeeded(lid, bid)
+	info.registered = true
+	return c.tryGenCert(ctx, logger, lid, bid, info)
 }
 
 // CertifyIfEligible signs the hare output, along with its role proof as a certifier, and gossip the CertifyMessage
 // if the node is eligible to be a certifier.
-func (c *Certifier) CertifyIfEligible(ctx context.Context, logger log.Log, lid types.LayerID, bid types.BlockID) error {
+func (c *Certifier) CertifyIfEligible(ctx context.Context, lid types.LayerID, bid types.BlockID) error {
 	if _, err := c.beacon.GetBeacon(lid.GetEpoch()); err != nil {
 		return errBeaconNotAvailable
 	}
-	// check if the node is eligible to certify the hare output
-	proof, err := c.oracle.Proof(ctx, lid, eligibility.CertifyRound)
+
+	var (
+		errsMu sync.Mutex
+		errs   error
+		eg     errgroup.Group
+	)
+	c.mu.Lock()
+	for _, s := range c.signers {
+		s := s
+		eg.Go(func() error {
+			if err := c.certifySingleSigner(ctx, s, lid, bid); err != nil {
```

Comment:
how about inlining certifySingleSigner here?

Reply:
I extracted it to simplify error handling.